File: leetcode/company-wise/google/0297_serialize_deserialize_bt.py
# LC 297. Serialize and Deserialize Binary Tree | Hard | Google
class TreeNode:
    def __init__(self, x): self.val = x; self.left = self.right = None

class Codec:
    def serialize(self, root):
        res = []
        def pre(node):
            if not node: res.append('N'); return
            res.append(str(node.val)); pre(node.left); pre(node.right)
        pre(root); return ','.join(res)

    def deserialize(self, data):
        vals = iter(data.split(','))
        def build():
            v = next(vals)
            if v == 'N': return None
            node = TreeNode(int(v)); node.left = build(); node.right = build()
            return node
        return build()

File: leetcode/company-wise/google/0297_serialize_deserialize_bt.cpp
// LC 297. Serialize and Deserialize Binary Tree | Hard | Google
#include <string>
#include <sstream>
#include <queue>
using namespace std;
struct TreeNode { int val; TreeNode *left, *right; TreeNode(int x) : val(x), left(nullptr), right(nullptr) {} };

class Codec {
public:
    string serialize(TreeNode* root) {
        if (!root) return "N";
        return to_string(root->val) + "," + serialize(root->left) + "," + serialize(root->right);
    }
    TreeNode* deserialize(string data) {
        istringstream ss(data); return build(ss);
    }
private:
    TreeNode* build(istringstream& ss) {
        string val; getline(ss, val, ',');
        if (val == "N") return nullptr;
        TreeNode* node = new TreeNode(stoi(val));
        node->left = build(ss); node->right = build(ss);
        return node;
    }
};
